### ISSUE TITLE
MSSQL - Class constructor ConnectionPool cannot be invoked without 'new'

### ISSUE THREAD

**Issue by luzzif** [knex/knex#3737]
Knex version: `0.20.11`
MSSQL version: `6.2.0`

So I'm trying to perform a union on 3 different datasets, and I keep on having this error (environment-specific paths removed):
```
TypeError: Class constructor ConnectionPool cannot be invoked without 'new'
    at MSSQL_Formatter.compileCallback (knex/lib/formatter.js:221:14)\n    
    at MSSQL_Formatter.parseObject (knex/lib/formatter.js:253:31)
    at MSSQL_Formatter.wrap (knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.columnize (knex/lib/formatter.js:77:19)
    at QueryCompiler_MSSQL.columns (knex/lib/dialects/mssql/query/compiler.js:147:35)
    at knex/lib/dialects/mssql/query/compiler.js:31:69
    at Array.map (<anonymous>)
    at QueryCompiler_MSSQL.select (knex/lib/dialects/mssql/query/compiler.js:31:35)
    at QueryCompiler_MSSQL.toSQL (knex/lib/query/compiler.js:64:29)
    at MSSQL_Formatter.compileCallback (knex/lib/formatter.js:228:21)
```

The offending code is the following:
```
const databaseClient = knex({...});
const a = databaseClient("table_a").select(...);
const b = databaseClient("table_b").select(...);
const c = databaseClient("table_c").select(...);

// Tried all of the following alternatives
const result = await a.unionAll(b).unionAll(c);
const result = await a.unionAll([b, c]);
const result = await a.unionAll(b.unionAll(c));
```

@smorey2

**Comment by briandamaged** [comment]
Howdy!  Is `6.2.0` referring to the version of the `mssql` package in node, or the version of the database server itself?

A quick glance at the `knex` code suggests that version `6.x` is not officially supported yet:

https://github.com/knex/knex/blob/31c5b86ec4ebadc1dfdd21afc6a3438ca24663b6/lib/dialects/mssql/index.js#L65-L76

Assuming that it's not currently supported: would you be interested in helping to add support for it?

**Comment by briandamaged** [comment]
Woops -- I might have completely misread the above code.  I saw the `^5.0.0` bit, but didn't notice that it was inside of a comment.  I guess this is why I should have my coffee before sitting in front of the computer.  lol

So, I'll try to take a look later today.

**Comment by luzzif** [comment]
Thank you very much for the fast reply! And yes, the MSSQL version specified above refers to the Node driver, not the database version itself.

**Comment by briandamaged** [comment]
@luzzif : Howdy again!  So, I started looking into this, but the line numbers in the stacktrace don't seem to line up w/ the source code.  Are you sure that you're using version `0.20.11` of `knex`?  You might want to run:

```
npm ls knex
```

to double-check it.

**Comment by luzzif** [comment]
I can confirm, running `npm ls knex` gives me back `knex@0.20.11`. I triple checked in the `node_modules` folder too, and `knex`'s `package.json` has the `0.20.11` version.

**Comment by luzzif** [comment]
News? It's pretty urgent.

**Comment by briandamaged** [comment]
So far, I have been unable to recreate the issue.  All of the references I see in the `knex` code base are using `new` when constructing a `ConnectionPool`.

Is there any possibility that:

1. Your `package.json` is obtaining `knex` from a different `git` repo?  (This would cause the version to be reported as `0.20.11`, but the underlying code might be different)
2. You are running some kind of bundling / linting tool that might be modifying the `knex` code as part of the build process?  (If so, then this would throw the line numbers off)

**Comment by luzzif** [comment]
Ok so, by looking at the lock file, the `knex` dependency is fetched from 
```
https://registry.yarnpkg.com/knex/-/knex0.20.11.tgz#c25167cdd562b08252a53b833b91cb6436e76a35
```

Which seems ok.

I am using Babel with `preset-env` to bundle my project.
But the weird thing is that the problem only appears when performing a union with 3 datasets... if I perform a "classic" double union (a union b), there's no problem at all!

**Comment by briandamaged** [comment]
k -- still working on recreating the issue, but it looks like you might also be invoking the `.unionAll(..)` method incorrectly.  Here's an example that I see from within the `knex` unit tests:

https://github.com/knex/knex/blob/e7e7a07ce90f73a29faabc6c094cd0b114c32622/test/unit/query/builder.js#L2097-L2113

Notice how 2 out of the 3 queries are established from within nested functions.  Additionally, they reference `this` rather than the top-level `knex` object.

**Comment by luzzif** [comment]
Just tried it. Same outcome, unfortunately...

**Comment by briandamaged** [comment]
Here's what I just tried:

```javascript

    context("hmm", function() {
      beforeEach(async function() {
        await knex.schema.dropTableIfExists('unionize');

        await knex.schema.createTable('unionize', function(t) {
          t.string('username');
        });
      });

      it.only('works', async function() {
        await knex('unionize').insert([
          {username: "batman"},
          {username: "robin"},
        ]);
        const a = knex("unionize").select("*");
        const b = knex("unionize").select("*");
        const c = knex("unionize").select("*");

        const result = await a.unionAll(b).unionAll(c);

        console.log(result);
      });

    });
```

It's running without any errors, and it's producing the output I was expecting to see.

Can you provide any further details about the queries that are being run?  Ex: perhaps one of them is creating a callback that bypasses `knex` and operates directly on the `mssql` / `tedious` libraries?

**Comment by briandamaged** [comment]
A few other things for you to try:

1. Run the queries individually.  Do any of them cause the same error when run alone?
2. Compare the stacktrace against the code you find in `node_modules/knex`.  Do the line numbers actually match up?
3. Confirm that the `acquireRawConnection(..)` method in your local copy of `knex` looks like the following:

https://github.com/knex/knex/blob/e7e7a07ce90f73a29faabc6c094cd0b114c32622/lib/dialects/mssql/index.js#L218-L234

4. Confirm that you don't have any code within your application that is overriding / patching methods in `knex`.  (Specifically, the `acquireRawConnection(..)` method)

**Comment by luzzif** [comment]
In the meantime, I just completely removed Babel from my project (it wasn't exactly pleasant ahah) and I tried to rerun the query. I got the exact same error with the following stack trace: 

```
TypeError: Class constructor ConnectionPool cannot be invoked without 'new'
    at MSSQL_Formatter.compileCallback (node_modules/knex/lib/formatter.js:221:14)
    at MSSQL_Formatter.parseObject (node_modules/knex/lib/formatter.js:253:31)
    at MSSQL_Formatter.wrap (node_modules/knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (node_modules/knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (node_modules/knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (node_modules/knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (node_modules/knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.parseObject (node_modules/knex/lib/formatter.js:265:27)
    at MSSQL_Formatter.wrap (node_modules/knex/lib/formatter.js:181:21)
    at MSSQL_Formatter.columnize (node_modules/knex/lib/formatter.js:77:19)
    at QueryCompiler_MSSQL.columns (node_modules/knex/lib/dialects/mssql/query/compiler.js:144:35)
    at node_modules/knex/lib/dialects/mssql/query/compiler.js:28:69
    at Array.map (<anonymous>)
    at QueryCompiler_MSSQL.select (node_modules/knex/lib/dialects/mssql/query/compiler.js:28:35)
    at QueryCompiler_MSSQL.toSQL (node_modules/knex/lib/query/compiler.js:64:29)
    at MSSQL_Formatter.compileCallback (node_modules/knex/lib/formatter.js:228:21)
```
At a first glance, it seems to be same as the one given above... so Babel probably isn't a problem...

As for your questions:
1. I tried running the three queries. I didn't get any results because of timeouts, but the mere fact that I got a timeout means that Knex correctly processed the queries themselves and sent them to the remote database.

2. I think so. Here's what I find at the first stack trace entry:
```
// Compiles a callback using the query builder.
  compileCallback(callback, method) {
    const { client } = this;

    // Build the callback
    const builder = client.queryBuilder();
    callback.call(builder, builder); // <- offending line is here!

    // Compile the callback, using the current formatter (to track all bindings).
    const compiler = client.queryCompiler(builder);
    compiler.formatter = this;

    // Return the compiled & parameterized sql.
    return compiler.toSQL(method || builder._method || 'select');
  }
```

3. I can confirm that the code matches.

4. Now that I removed Babel, I have nothing else apart from the linter which could cause trouble, but the linter doesn't run on `node_modules`, so I think it's pretty safe to say I don't have anything overriding any dependency at all

**Comment by luzzif** [comment]
I got it right after relooking at the code... I just wrapped in a function the query builder passed to the `unionAll`, and I can now see the query on screen.

Now I've got another error, but I don't think it's related...

**Comment by briandamaged** [comment]
Ah -- so, there are several different reasons that a `TimeoutError` / `KnexTimeoutError` might get thrown.  In some cases, the query will never actually reach the database.  (Ex: if you run out of connection).

So, you might need to check the database logs to confirm that the query actually reached the DB.

**Comment by luzzif** [comment]
Ok, there's still something wrong. The query is generated, but the second and third `unionAll` datasets are not really there.

Example of generated query:

```
SELECT * FROM foo WHERE ... UNION ALL (SELECT *) UNION ALL (SELECT *)
```

The way in which I now generate the query is the following:

```
const databaseClient = knex({ ... });
const foo = databaseClient.select({ ... }).from("foo");
const bar = databaseClient.select({ ... }).from("bar");
const joe = databaseClient.select({ ... }).from("joe");

foo.unionAll([() => bar, () => receipts]);
```

Another thing I didn't tell you: this union query is a third-level query (i.e. `SELECT * FROM (SELECT * FROM (MY_QUERY_HERE))`).

**Comment by luzzif** [comment]
I finally solved it, oh my God...

The third query in the `unionAll` had an error... I was using `knex.raw("")` to pick an empty value from db, plus two other `knex.raw(variableContainingString)` without using any single quotes at all. Once I added all the missing single quotes, it worked like a charm.

Thank you for the support anyway, I appreciate it!

**Comment by kibertoad** [comment]
@luzzif Wonder if our error messages could be more helpful in this case, though. Can you take a look at the call chain in your case and see if there are easy fail-fast wins to have there?

**Comment by luzzif** [comment]
One weird thing is that I got the apparently same syntax error in two different places (the `raw` without single quotes), but with two different outcomes.

In the first case I got that cryptic error, while in the second one I got a pretty good error message, which immediately would have made me think about the correct fix to put in place (the lucky dev I am, I only first encountered the first error, and noticed the second only after already having fixed them all).

The only difference that I noticed between the two cases is in the two queries nesting. 

As I said above, the first was a third-level query, while the second one a second-level query.
